server/bin/pbench-unpack-tarballs.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+import os
+from pathlib import Path
+import sys
+
+from pbench.common.exceptions import BadConfig
+from pbench.common.logger import get_pbench_logger
+from pbench.server import PbenchServerConfig
+from pbench.server.database import init_db
+from pbench.server.unpack_tarballs import UnpackTarballs
+
+_NAME_ = "pbench-unpack-tarballs"
+
+# The link source for the operation of this script.
+_linksrc = "TO-UNPACK"
+
+
+def main(cfg_name):
+    if not cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set"
+            " _PBENCH_SERVER_CONFIG env variable.",
+            file=sys.stderr,
+        )
+        return 2
+
+    try:
+        config = PbenchServerConfig(cfg_name)
+    except BadConfig as e:
+        print(f"{_NAME_}: {e}", file=sys.stderr)
+        return 1
+
+    logger = get_pbench_logger(_NAME_, config)
+
+    # We're going to need the Postgres DB to track dataset state, so setup
+    # DB access.
+    init_db(config, logger)
+
+    try:
+        BUCKET = str(sys.argv[1])
+    except Exception as exc:
+        logger.error("Bucket not Defined: ", exc)
+        return 1
+
+    prog = Path(sys.argv[0]).name
+
+    lowerbound, upperbound = 0, float("inf")
+
+    if BUCKET:
+        lowerbound = config.get(f"pbench-unpack-tarballs/{BUCKET}", "lowerbound")
+        if lowerbound:
+            lowerbound *= 1024 * 1024
+        upperbound = config.get(f"pbench-unpack-tarballs/{BUCKET}", "upperbound")
+        if upperbound:
+            upperbound *= 1024 * 1024
```

Comment:
In fact, this is a bug. If we get a falsey value from the config file, we can't set `upperbound` to that value... we should leave it at the `float("inf")` default we set above. So how about

```
   u = config.get(f"pbench-unpack-tarballs/{BUCKET}", "upperbound")
   if u:
      upperbound = float(u) *= 1024.0 * 1024.0
```

Reply:
In theory, a site can override the `pbench-server-default.cfg` values, and we have a Jira somewhere about being cautious with the values we get. As I've mentioned elsewhere I'd rather drop this sort of configurable stuff entirely and remove the `.cfg` files in favor of just a few container environment variables... but that's outside the scope of this PR.

In any case, our current default `pbench-server-default.cfg` omits an `upperbound` for the `pbench-unpack-tarballs/huge` section to take the default of "infinity". (Note that in the old bash implementation, this translates to just omitting the `-size -<x>` term from the `find` command to limit the upper bound, so the actual value was irrelevant.) So for the `bucket == "huge"` instance we *will* (normally) get `None` here, not something we can interpret as a number.

Reply:
I'm not 💯 on this, but I _think_ we can specify a fallback value to be provided when the config value is missing!

If that's true, then perhaps we can handle things more gracefully:
```
    	MIN_LOWERBOUND = 0
        MAX_UPPERBOUND = float("inf")
...
        lowerbound, upperbound = MIN_LOWERBOUND, MAX_UPPERBOUND
...
        lowerbound = config.get(
            f"pbench-unpack-tarballs/{BUCKET}", "lowerbound", fallback=MIN_LOWERBOUND
        )
        if lowerbound:
            lowerbound *= 1024 * 1024

        upperbound = config.get(
            f"pbench-unpack-tarballs/{BUCKET}", "upperbound", fallback=MAX_UPPERBOUND
        )
	if upperbound:
            upperbound *= 1024 * 1024

```

Reply:
Huh. You know, `PbenchServerConfig` only contains a `ConfigParser` rather than extending it... but you're right; `PbenchConfig.get` passed through `kwargs` to the `ConfigParser.get` and `fallback=` ought to work.

However, if we're doing it that way then the `if lowerbound` and `if upperbound` serve no purpose except to avoid multiplying by 0... and we'd still be multiplying huge's `float("inf")` by `1024 * 1024` to get a **really big** ♾️.

Reply:
> the `if lowerbound` and `if upperbound` serve no purpose

Um, yeah.

However, I'm still nervous about `config.get()` returning something which isn't a number.  😬

> we'd still be multiplying huge's `float("inf")` by `1024 * 1024` to get a really big ♾️.

Yeah...I checked to make sure that that would work before I proposed it.  😁

Reply:
> > the `if lowerbound` and `if upperbound` serve no purpose
> 
> Um, yeah.
> 
> However, I'm still nervous about `config.get()` returning something which isn't a number. grimacing

Fair enough, but these conditionals only help if it's **falsey** garbage. If we're going to worry about non-numeric values, let's just catch `ValueError` and log it.

(And remember that this is intended to be run by `cron`, and the command line was generated at install time and ain't gonna change on its own so we'll be repeating this log entry every cron interval...)

Reply:
> these conditionals only help if it's *falsey* garbage.

Yeah, which is why I'm nervous!  🤣

> If we're going to worry about non-numeric values, let's just catch ValueError and log it.

Actually, given the conditions you describe, I don't really mind a crash (e.g., that's probably better than having it continue to run using default values...).  But, what I don't want is for `config.get()` to return a number _as a string_ and then have the `*` operator replicate it a million times and then have Python's duck typing try to turn it into a really huge, non-infinite, number....  😕

Reply:
Actually, that's a good point: `ConfigParser` isn't *typed*, so the result is always a string and we're depending on implicit conversion.

So I'll re-state my earlier proposal for an explicit `float(value)` (or `int(value)` for `lowerbound`); we'll get an explicit `ValueError` if it's not a string representation of a floating point number.

The `if lowerbound` is just a trivial subset of that check and provides little value. (And with `fallback=` we'll never get a `None` anyway.)

Reply:
> `ConfigParser` isn't typed, so the result is always a string and we're depending on implicit conversion.

In my limited, trivial, but obvious testing, I don't see any implicit conversion -- I get a `TypeError` when I try to treat a string as a number.

> I'll re-state my earlier proposal for an explicit `float(value)` (or `int(value)` for lowerbound)

I don't thing we have any other option but to do a forced conversion.  (Curiously, a quick search of our source doesn't show any other code which explicitly converts the return from `get()` to a number, although there are a few coercions to a `str()`....  🤔)  Since we're going to end up being explicit, I'll vote for `float()`.  That is, since the units are megabytes, it might be too coarse to limit it to integral numbers, so enabling values like `"1.5"` might be handy (and doing so won't rule out values like `"1"`).

So, yeah, lets switch to `float`'s.

Reply:
```
        MIN_LOWERBOUND = 0.0
        MAX_UPPERBOUND = float("inf")
...
        lowerbound, upperbound = MIN_LOWERBOUND, MAX_UPPERBOUND
...
            lowerbound = config.get(
                f"pbench-unpack-tarballs/{BUCKET}",
                "lowerbound",
                fallback=MIN_LOWERBOUND,
            )
            lowerbound = float(lowerbound) * 1024 * 1024

            upperbound = config.get(
                f"pbench-unpack-tarballs/{BUCKET}",
                "upperbound",
                fallback=MAX_UPPERBOUND,
            )
            upperbound = float(upperbound) * 1024 * 1024
```
(Silly Black....)

(BTW, it turns out that `float(u) *= 1024` is not legal Python syntax.)

And, FWIW, I might be inclined to move the assignments of the defaults into an `else` clause....

Reply:
> (BTW, it turns out that float(u) *= 1024 is not legal Python syntax.)

Ah, yes; one of those annoyingly picky languages that wants *variables* on the left hand side of assignments. 😆

Reply:
>  one of those annoyingly picky languages that wants variables on the left hand side of assignments. 😆

Actually, I think Python's requirements are much pickier and more arcane than just _variables_....  (I'm not even sure that Python _has_ _variables_...I think they are just symbolic references which get dynamically bound to things....)  🥴

Reply:
Yeah, sure, if we're going all pedantic here... Technically the left hand side is a "target list", which may be enclosed in parenthesis, one target per expression on the right-hand side. The simplest "target" is an "identifier", but of course we have list indices, object attributes, and other factors.

https://docs.python.org/3/reference/simple_stmts.html#assignment-statements

Reply:
It was a good Discussion arounf ConfigParser